### PR TITLE
FPS metric using the game windowed FPS

### PR DESCRIPTION
**Pull request by improbable-danny**

### CODE REVIEW

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.h [hunk 1]
```diff
@@ -35,6 +35,7 @@ class GDKTESTGYMS_API ABenchmarkGymGameMode : public AGameModeBase
 	double GetClientRTT() const { return AveragedClientRTTSeconds; }
 	double GetClientViewLateness() const { return AveragedClientViewLatenessSeconds; }
 	double GetPlayersConnected() const { return ActivePlayers; }
+	double GetFPSValid() const { return !bHasFpsFailed ? 1.0 : 0.0; }
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.h:38]: What sets bHasFpsFailed?

Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.h:38]: This line here https://github.com/spatialos/UnrealGDKTestGyms/pull/100/files#diff-0fc3b032552e5b3294cf5ce6fb65fde4R289

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 2]
```diff
@@ -15,6 +15,7 @@
 #include "Misc/CommandLine.h"
 #include "Misc/Crc.h"
 #include "Utils/SpatialMetrics.h"
+#include "NFRConstants.h"
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:18]: Sort these pls

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 3]
```diff
@@ -24,6 +25,8 @@ namespace
 	const FString AverageClientRTTMetricName = TEXT("UnrealAverageClientRTT");
 	const FString AverageClientViewLatenessMetricName = TEXT("UnrealAverageClientViewLateness");
 	const FString PlayersSpawnedMetricName = TEXT("UnrealActivePlayers");
+	const FString AverageFPSValid = TEXT("UnrealFPSValid");
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:28]: add server, useful for indicating that this will be testing the offloaded server in multi-worker also

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 4]
```diff
@@ -284,6 +293,17 @@ void ABenchmarkGymGameMode::ServerUpdateNFRTestMetrics(float DeltaSeconds)
 			}
 		}
 	}
+
+	if (!bHasClientFpsFailed)
+	{
+		if (!bClientFpsWasValid)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:299]: nit - single if

File: Game/Source/GDKTestGyms/Public/UserExperienceReporter.h [hunk 5]
```diff
@@ -23,13 +23,15 @@ class GDKTESTGYMS_API UUserExperienceReporter : public UActorComponent
...
 	float ServerViewLateness;
+	bool bFrameRateValid;
 
 	void InitializeComponent() override;
 	 
 	void OnClientOwnershipGained() override;
 	void ReportMetrics();
 	UFUNCTION(Server, reliable)
-	void ServerReportedMetrics(float RTTSeconds, float ViewLatenessSeconds);
+	void ServerReportedMetrics(float RTTSeconds, float ViewLatenessSeconds, bool bFrameRateValid);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/UserExperienceReporter.h:36]:
```suggestion
	void ServerReportedMetrics(float RTTSeconds, float ViewLatenessSeconds, bool bInFrameRateValid);
```

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 6]
```diff
@@ -97,6 +97,18 @@ void ABenchmarkGymGameMode::BeginPlay()
 				Delegate.BindUObject(this, &ABenchmarkGymGameMode::GetPlayersConnected);
 				SpatialDriver->SpatialMetrics->SetCustomMetric(PlayersSpawnedMetricName, Delegate);
 			}
+
+			{
+				UserSuppliedMetric Delegate;
+				Delegate.BindUObject(this, &ABenchmarkGymGameMode::GetFPSValid);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:103]: Driveby - but worth doing now - can we only register this delegate for all gamemode - and register the other 4 if you're authoritative?

File: Game/Source/GDKTestGyms/Public/NFRConstants.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#pragma once
+
+#include "CoreMinimal.h"
+
+DECLARE_LOG_CATEGORY_EXTERN(LogNFRConstants, Log, All);
+
+class NFRConstants
+{
+	int64 TimeToStartFPSSampling;
+	bool bFPSSamplingValid{false};
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/NFRConstants.h:12]: Even spacing

File: Game/Source/GDKTestGyms/Public/NFRConstants.h [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	bool bInitialized{ false };
+	float MinServerFPS = 20.0f;
+	float MinClientFPS = 20.0f;
+
+	NFRConstants() {}
+public:
+	
+	bool SamplesForFPSValid(); // Wait until this is true until checking FPS values
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/NFRConstants.h:20]: uber nit: can we put public section at the top, order the functions according to use (singleton getter first) etc. then add the private section for methods and properties.

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+			MinClientFPS = FCString::Atof(*MinClientFPSStr);
+		}
+	}
+	UE_LOG(LogNFRConstants, Log, TEXT("Min client FPS: %.8f."), MinClientFPS);
+	bInitialized = true;
+}
+
+NFRConstants& NFRConstants::Get()
+{
+	static NFRConstants _;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/NFRConstants.cpp:63]: you've gone mad with power danny

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp [hunk 10]
```diff
@@ -0,0 +1,77 @@
+#include "NFRConstants.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "Interop/SpatialWorkerFlags.h"
+
+DEFINE_LOG_CATEGORY(LogNFRConstants);
+
+bool NFRConstants::SamplesForFPSValid()
+{
+	if (bFPSSamplingValid)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/NFRConstants.cpp:9]: Add check(bInitialized);  ?

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	return bIsValidNow;
+}
+
+void NFRConstants::Init(UWorld* World)
+{
+	TimeToStartFPSSampling = FDateTime::Now().GetTicks() + FTimespan::FromSeconds(120.0f).GetTicks();
+
+	USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(World->GetNetDriver());
+	// Sever fps
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/NFRConstants.cpp:26]: sever that fps!

File: Game/Source/GDKTestGyms/GDKTestGymsGameInstance.cpp [hunk 12]
```diff
@@ -15,6 +16,7 @@ void UGDKTestGymsGameInstance::Init()
 	TickDelegateHandle = FTicker::GetCoreTicker().AddTicker(TickDelegate);
 	TickWindowTotal = 0;
 	GetEngine()->NetworkFailureEvent.AddUObject(this, &UGDKTestGymsGameInstance::NetworkFailureEventCallback);
+	NFRConstants::Get().Init(GetWorld());
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/GDKTestGymsGameInstance.cpp:19]: I like aspects of this - but I feel it might fall down when testing in editor. Thoughts?

Comment (improbable-danny) [reply Game/Source/GDKTestGyms/GDKTestGymsGameInstance.cpp:19]: Yeah possibly, maybe its worth creating an actor for this like we do for latency tracer

Comment (improbable-danny) [reply Game/Source/GDKTestGyms/GDKTestGymsGameInstance.cpp:19]: On second thoughts, changing worker flags in the editor will for sure break this

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp (hunk 9 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/Private/NFRConstants.cpp:63]: I cannot be stopped

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp (hunk 10 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/NFRConstants.cpp:17]: bump

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	return bIsValidNow;
+}
+
+void UNFRConstants::InitWithWorld(UWorld* World)
+{
+	TimeToStartFPSSampling = FDateTime::Now().GetTicks() + FTimespan::FromSeconds(120.0f).GetTicks();
+
+	USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(World->GetNetDriver());
+	// Server fps
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/NFRConstants.cpp:34]: These overrides aren't actually used anywhere yet right?

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp [hunk 14]
```diff
@@ -7,6 +7,8 @@
 #include "SpatialNetDriver.h"
 #include "UserExperienceComponent.h"
 #include "Utils/SpatialMetrics.h"
+#include "GDKTestGymsGameInstance.h"
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp:10]: Sort these

File: Game/Source/GDKTestGyms/Public/NFRConstants.h [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+	float GetMinClientFPS() const;
+	
+	static const UNFRConstants* Get(UWorld* World);
+private:
+	int64 TimeToStartFPSSampling;
+	mutable bool bFPSSamplingValid{ false };
+	float MinServerFPS = 20.0f;
+	float MinClientFPS = 20.0f;
+
+	UNFRConstants();
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/NFRConstants.h:31]: Do we actually need the constructor?

File: Game/Source/GDKTestGyms/Private/NFRConstants.cpp (hunk 13 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/Private/NFRConstants.cpp:35]: They aren't but just so they are available for when we want to (latency test)
